### PR TITLE
*: sort by ASCEND on missing sort order

### PR DESCRIPTION
**Pull request by gyuho**
Address https://github.com/coreos/etcd/issues/6671.

/cc @xiang90

### CODE REVIEW

File: etcdctl/ctlv3/command/get_command.go [hunk 1]
```diff
@@ -105,6 +106,13 @@ func getGetOp(cmd *cobra.Command, args []string) (string, []clientv3.OpOption) {
 		opts = append(opts, clientv3.WithRev(getRev))
 	}
 
+	if getSortOrder == "" && getSortTarget != "" {
+		ExitWithError(ExitBadFeature, errors.New("missing sort order (--order)"))
```
Comment (xiang90) [inline etcdctl/ctlv3/command/get_command.go]: exitbadflag?

Comment (gyuho) [inline etcdctl/ctlv3/command/get_command.go]: We don't have bad flag exit. Instead used `ExitBadArgs`. Is that ok?

PTAL.

**Comment by xiang90** [comment]
If there is no order given, shouldnt etcd assumes an ordering?

**Comment by xiang90** [comment]
I do not know why i tell etcdctl to sort by create revision but it does not.

**Comment by gyuho** [comment]
@xiang90 `WithSort` does

```go
func WithSort(target SortTarget, order SortOrder) OpOption {
	return func(op *Op) {
		if target == SortByKey && order == SortAscend {
			// If order != SortNone, server fetches the entire key-space,
			// and then applies the sort and limit, if provided.
			// Since current mvcc.Range implementation returns results
			// sorted by keys in lexiographically ascending order,
			// client should ignore SortOrder if the target is SortByKey.
			order = SortNone
		}
		op.sort = &SortOption{target, order}
	}
}
```

But this doesn't cover the case of `sort-by=CREATE`.

It works when `sort-by=KEY` because it returns lexiographically ascending order by default.

**Comment by xiang90** [comment]
We should fill in a default order for users when they specify sort-by x when x is not key.

**Comment by gyuho** [comment]
@xiang90 Right that's better. Will rework on this with more tests. Thanks.

**Comment by gyuho** [comment]
@xiang90 For users, it's now `ASCEND by default` on all targets.

Internally only run sorting when the target is not key and order is not ascend.

PTAL.

File: etcdctl/ctlv3/command/get_command.go [hunk 2]
```diff
@@ -137,6 +137,13 @@ func getGetOp(cmd *cobra.Command, args []string) (string, []clientv3.OpOption) {
 		ExitWithError(ExitBadFeature, fmt.Errorf("bad sort target %v", getSortTarget))
 	}
 
+	if sortByTarget != clientv3.SortByKey && sortByOrder == clientv3.SortNone {
+		// Since current mvcc.Range implementation returns results
```
Comment (xiang90) [inline etcdctl/ctlv3/command/get_command.go]: alternatively, can we fix this at server side? in apply.go line 321?

Comment (xiang90) [inline etcdctl/ctlv3/command/get_command.go]: if any target is specified and sort_order == none, we should set sort order to ascend.

Comment (gyuho) [inline etcdctl/ctlv3/command/get_command.go]: Sure will try that.

**Comment by gyuho** [comment]
@xiang90 Moved the change to server side and added more tests and docs.

PTAL.

Thanks.

File: clientv3/op.go [hunk 3]
```diff
@@ -227,6 +227,8 @@ func WithSort(target SortTarget, order SortOrder) OpOption {
 			// client should ignore SortOrder if the target is SortByKey.
 			order = SortNone
 		}
+		// server-side sets order == SortAscend by default
```
Comment (xiang90) [inline clientv3/op.go]: remove this comment

Comment (gyuho) [inline clientv3/op.go]: ok removed.

File: etcdserver/apply.go [hunk 4]
```diff
@@ -323,6 +323,11 @@ func (a *applierV3backend) Range(txnID int64, r *pb.RangeRequest) (*pb.RangeResp
 			sort.Sort(sorter)
 		case r.SortOrder == pb.RangeRequest_DESCEND:
 			sort.Sort(sort.Reverse(sorter))
+		case r.SortTarget != pb.RangeRequest_KEY:
```
Comment (xiang90) [inline etcdserver/apply.go]: i do not understand this. we will only reach this code if sort order is not none (see 307). What we want is to sort the keys if target == create and order == none.

Comment (gyuho) [inline etcdserver/apply.go]: Right, will fix.

**Comment by gyuho** [comment]
@xiang90 PTAL. Now it properly sets `ASCEND` when target is not key and sort order is none.

Thanks!

File: clientv3/integration/kv_test.go [hunk 5]
```diff
@@ -226,6 +226,21 @@ func TestKVRange(t *testing.T) {
 				{Key: []byte("fop"), Value: nil, CreateRevision: 9, ModRevision: 9, Version: 1},
 			},
 		},
+		// range all with SortByKey, missing sorting order (ASCEND by default)
```
Comment (xiang90) [inline clientv3/integration/kv_test.go]: we probably should also add a server side test.

**Comment by xiang90** [comment]
LGTM

**Comment by gyuho** [comment]
@xiang90  Also added a test case in server side. Will merge if there's no objection? Thanks.

**Comment by xiang90** [comment]
LGTM

**Pull request merged**